### PR TITLE
WebGLRenderer: Support "Linear Display P3" working color space and "Display P3" unlit rendering

### PR DESCRIPTION
**Pull request by donmccurdy**
**tl;dr** — Adds support for wide gamut color spaces in unlit rendering workflows such as 2D games, photogrammetry, and image processing applications.

***

Extends THREE.ColorManagement and THREE.WebGLRenderer with support — only in "unlit" scenes, currently — for the wide-gamut Display P3 color space. To use this feature, the following steps are required:

1. Assign `ColorManagement.workingColorSpace = LinearDisplayP3ColorSpace`
2. Assign `renderer.outputColorSpace = DisplayP3ColorSpace`
3. Assign `texture.colorSpace = DisplayP3ColorSpace` AND ensure each texture's ICC Profile (or other color space metadata) is correctly embedded

To see the difference in this example, you'll need use Chrome or Safari, and a display and operating system that support the Display P3 wide gamut color space. Most newer macOS devices will do fine, but your external monitor may not. The change covers only unlit rendering now, e.g. MeshBasicMaterial. More changes will be required to adjust the lit rendering pipeline for a wider gamut, and @WestLangley and I are still discussing those requirements.

***

~~The images attached are borrowed from https://webkit.org/blog-files/color-gamut/comparison.html. We will want to verify the licenses and attribution before merging this PR, or create our own examples.~~

***

Related

- https://github.com/mrdoob/three.js/issues/26479

### CODE REVIEW

File: src/math/ColorManagement.js [hunk 1]
```diff
@@ -105,29 +106,67 @@ export const ColorManagement = {
...
+			ctx.drawingBufferColorSpace = colorSpace;
+			return ctx.drawingBufferColorSpace === colorSpace; // deepscan-disable-line SAME_OPERAND_VALUE
+
+		} catch ( e ) {
+
+			return false;
+
+		}
+
+	}
```
Comment (donmccurdy) [inline src/math/ColorManagement.js:157]: I'm not sure `isDrawingBufferColorSpaceSupported` belongs in ColorManagement... but ideally we need to be able to check this as early as possible... Ideas would be welcome!

Comment (donmccurdy) [reply src/math/ColorManagement.js:157]: I'm leaning toward doing this, instead:

```javascript
if ( WebGL.isColorSpaceAvailable( THREE.DisplayP3ColorSpace ) ) {

  THREE.ColorManagement.workingColorSpace = THREE.LinearDisplayP3ColorSpace;

}
```

Comment (Mugen87) [reply src/math/ColorManagement.js:157]: LGTM!

**Comment by donmccurdy** [comment]
To briefly summarize a few things happening in this PR and its demo —

1. When the _browser_ supports Display P3 (In 2023: Chrome and Safari), and the developer knows their scene contains wide gamut content, we assign `ColorManagement.workingColorSpace = LinearDisplayP3ColorSpace` to enable wide gamut _rendering_.
2. When the _display_ supports Display P3 (In 2023: Most macOS, iOS, and Pixel devices) we assign `renderer.outputColorSpace = DisplayP3ColorSpace` to enable wide gamut _output_. This may change during a session, if displays are (dis)connected or windows are moved to another display.
3. When a texture's `.colorSpace` property indicates that it uses primaries differing from those of the working color space (e.g. sRGB textures in a Linear Display P3 working color space) the renderer enables `gl.UNPACK_COLORSPACE_CONVERSION_WEBGL` to gamut map the texture to the working color space. This operation requires that the texture's ICC profile be correct; setting `texture.colorSpace` alone is not enough. WebGL cannot do this for compressed textures or data textures — they must already use the working primaries.
4. When a texture's `.colorSpace` property indicates that it uses a supported transfer function (both sRGB and Display P3 use the "sRGB" transfer function) the renderer chooses an internal format such that WebGL will automatically apply the sRGB EOTF when sampling from the texture.
5. After rendering, pixels are encoded to the output color space.

As a result of (3) and (4), color images are typically uploaded to WebGL in either sRGB or Display P3 (whichever is the non-linear analog to the linear working color space). When the GPU samples the textures, it decodes each sample by applying the sRGB EOTF, to obtain a sample in the working color space (Linear sRGB or Linear Display P3).

**Comment by donmccurdy** [comment]
I've simplified the demo, using just a three.js logo. Copyright assigned to the three.js authors, Creative Commons Attribution 3.0. This is ready to merge experimentally, when we are ready. We can then continue working on changes required for lit rendering.

File: src/renderers/webgl/WebGLBackground.js [hunk 2]
```diff
@@ -108,7 +109,7 @@ function WebGLBackground( renderer, cubemaps, cubeuvmaps, state, objects, alpha,
 			boxMesh.material.uniforms.flipEnvMap.value = ( background.isCubeTexture && background.isRenderTargetTexture === false ) ? - 1 : 1;
 			boxMesh.material.uniforms.backgroundBlurriness.value = scene.backgroundBlurriness;
 			boxMesh.material.uniforms.backgroundIntensity.value = scene.backgroundIntensity;
-			boxMesh.material.toneMapped = ( background.colorSpace === SRGBColorSpace ) ? false : true;
+			boxMesh.material.toneMapped = ( ColorManagement.getTransfer( background.colorSpace ) === SRGBTransfer ) ? false : true;
```
Comment (LeviPesin) [inline src/renderers/webgl/WebGLBackground.js:112]:
```suggestion
			boxMesh.material.toneMapped = ColorManagement.getTransfer( background.colorSpace ) !== SRGBTransfer;
```

File: src/renderers/webgl/WebGLBackground.js [hunk 3]
```diff
@@ -164,7 +165,7 @@ function WebGLBackground( renderer, cubemaps, cubeuvmaps, state, objects, alpha,
 
 			planeMesh.material.uniforms.t2D.value = background;
 			planeMesh.material.uniforms.backgroundIntensity.value = scene.backgroundIntensity;
-			planeMesh.material.toneMapped = ( background.colorSpace === SRGBColorSpace ) ? false : true;
+			planeMesh.material.toneMapped = ( ColorManagement.getTransfer( background.colorSpace ) === SRGBTransfer ) ? false : true;
```
Comment (LeviPesin) [inline src/renderers/webgl/WebGLBackground.js:168]:
```suggestion
			planeMesh.material.toneMapped = ColorManagement.getTransfer( background.colorSpace ) !== SRGBTransfer;
```

File: src/renderers/WebGLRenderer.js [hunk 4]
```diff
@@ -2392,6 +2395,13 @@ class WebGLRenderer {
 
 		};
 
+		this.resetColorManagement = function () {
+
+			_gl.drawingBufferColorSpace = this._outputColorSpace === DisplayP3ColorSpace ? 'display-p3' : 'srgb';
+			_gl.unpackColorSpace = ColorManagement.workingColorSpace === LinearDisplayP3ColorSpace ? 'display-p3' : 'srgb';
+
+		};
+
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js:2404]: Do we need this new method?
Could we inline the calls in `set outputColorSpace( colorSpace )`?

Comment (donmccurdy) [reply src/renderers/WebGLRenderer.js]: No need for it – or we don't need a public method, at least. I've inlined the calls in the setter. ✅

**Comment by donmccurdy** [comment]
FYI, this demo has just broken in Safari on macOS Sonoma:

https://bugs.webkit.org/show_bug.cgi?id=262429
